File: Cpp/Example6-1-DemReader/Example6-1-DemReader.cpp
// Example1DemReader.cpp : 此文件包含 "main" 函数。程序执行将在此处开始并结束。
//

#include <gdal_priv.h>

#include <iostream>

using namespace std;

int main() {
  GDALAllRegister();  //注册格式

  //设置Proj数据
  std::string projDataPath = getenv("GISBasic");
  projDataPath += "/share/proj";
  CPLSetConfigOption("PROJ_LIB", projDataPath.c_str());

  string workDir = getenv("GISBasic");
  string demPath = workDir + "/../Data/Terrain/dem.tif";

  GDALDataset* dem = (GDALDataset*)GDALOpen(demPath.c_str(), GA_ReadOnly);
  if (!dem) {
    cout << "Can't Open Image!" << endl;
    return 1;
  }

  int demWidth = dem->GetRasterXSize();
  int demHeight = dem->GetRasterYSize();
  int bandNum = dem->GetRasterCount();

  cout << "DEM列数：" << demWidth << endl;
  cout << "DEM行数：" << demHeight << endl;
  cout << "数据集波段数：" << bandNum << endl;

  const char* stringDataType =
      GDALGetDataTypeName(dem->GetRasterBand(1)->GetRasterDataType());
  cout << "DEM数据类型：" << stringDataType << endl;

  int success = 0;
  cout << "DEM无效值：" << dem->GetRasterBand(1)->GetNoDataValue(&success)
       << endl;

  cout << "DEM空间参考坐标系：\n";
  const char* wktString = dem->GetProjectionRef();
  cout << wktString << endl;

  //坐标信息
  double geoTransform[6] = {0};
  dem->GetGeoTransform(geoTransform);
  double dx = geoTransform[1];
  double dy = geoTransform[5];
  double startX = geoTransform[0] + 0.5 * dx;
  double startY = geoTransform[3] + 0.5 * dy;

  cout << "DEM间距：" << dx << '\t' << dy << '\n';
  cout << fixed << "DEM左上角起点位置：" << startX << '\t' << startY << endl;

  cout << "读取DEM高程..." << endl;
  size_t demBufNum = (size_t)demWidth * demHeight;
  vector<float> demBuf(demBufNum, 0);

  int depth = sizeof(float);
  dem->GetRasterBand(1)->RasterIO(GF_Read, 0, 0, demWidth, demHeight,
                                  demBuf.data(), demWidth, demHeight,
                                  GDT_Float32, depth, demWidth * depth);
  cout << "完成" << endl;

  GDALClose(dem);
}

File: Python/example-6-1-dem-reader.py
import os
import sys
from osgeo import gdal
from osgeo import osr
import numpy as np

# 自动设置 PROJ_LIB for Conda (Windows/Linux/macOS) 
def setup_proj_lib():
    if 'CONDA_PREFIX' in os.environ:
        conda_prefix = os.environ['CONDA_PREFIX']
        # Windows
        proj_path = os.path.join(conda_prefix, 'Library', 'share', 'proj')
        if not os.path.exists(proj_path):
            # Unix-like
            proj_path = os.path.join(conda_prefix, 'share', 'proj')
        if os.path.exists(proj_path):       
            os.environ['PROJ_LIB'] = proj_path
            return True
    return False

def main():
    # 注册所有 GDAL 支持的格式驱动
    gdal.AllRegister()

    # 拼接 DEM 文件路径
    gis_basic = os.getenv("GISBasicRepo")
    if not gis_basic:
        print("错误：环境变量 GISBasicRepo 未设置。", file=sys.stderr)
        return 1
    dem_path = os.path.join(gis_basic, "Data", "Terrain", "dem.tif")

    # 以只读模式打开 DEM 数据集
    dem = gdal.Open(dem_path, gdal.GA_ReadOnly)
    if not dem:
        print("Can't Open Image!")
        return

    # 获取 DEM 的基本信息
    dem_width = dem.RasterXSize
    dem_height = dem.RasterYSize
    band_num = dem.RasterCount

    print(f"DEM列数：{dem_width}")
    print(f"DEM行数：{dem_height}")
    print(f"数据集波段数：{band_num}")

    # 获取第一个波段的数据类型名称
    band1 = dem.GetRasterBand(1)
    string_data_type = gdal.GetDataTypeName(band1.DataType)
    print(f"DEM数据类型：{string_data_type}")

    # 获取无效值 (NoData Value)
    no_data_value = band1.GetNoDataValue()
    if no_data_value is None:
        print("DEM无效值：None (未设置)")
    else:
        print(f"DEM无效值：{no_data_value}")

    # 获取空间参考坐标系 (WKT字符串)
    print("DEM空间参考坐标系：")
    wkt_string = dem.GetProjection()
    print(wkt_string)

    # 获取地理变换参数（坐标信息）
    geo_transform = dem.GetGeoTransform()
    # geo_transform 包含 [起点X, 像素宽, 旋转参数, 起点Y, 旋转参数, 像素高]
    dx = geo_transform[1]
    dy = geo_transform[5]
    # 计算左上角第一个像素的中心点坐标
    start_x = geo_transform[0] + 0.5 * dx
    start_y = geo_transform[3] + 0.5 * dy

    print(f"DEM间距：{dx}\t{dy}")
    print(f"DEM左上角起点位置：{start_x}\t{start_y}")

    # 读取 DEM 高程数据
    print("读取DEM高程...")
    # 直接将数据读取为 NumPy 数组 (对应 C++ 的 vector<float>)
    # ReadAsArray 默认会读取为 float32 (如果源数据是 float32)
    dem_buf = band1.ReadAsArray(0, 0, dem_width, dem_height).astype(np.float32)
    print("完成")
    
    # 打印前几个高程值作为验证（可选）
    # print("前5个高程值:", dem_buf.flatten()[:5])

    # 释放数据集对象
    dem = None

if __name__ == "__main__":
    osr.UseExceptions() # 启用异常处理（推荐），避免静默错误
    setup_proj_lib() #设置 PROJ_LIB 
    main()